api/models/datastore.go

```diff
@@ -5,21 +5,66 @@ import (
+// - leading slash
+// - overall legal/scrubbed path (no relative paths, consecutive slashes, etc)
+// - trailing slash allowed and distinct from none
 type Datastore interface {
-	// GetApp returns the app called appName or nil if it doesn't exist
+
+	// GetApp gets an App by name.
+	// Returns ErrDatastoreEmptyAppName for empty appName.
+	// Returns ErrAppsNotFound if no app is found.
 	GetApp(ctx context.Context, appName string) (*App, error)
+
+	// GetApps gets a slice of Apps, optionally filtered by name.
+	// Missing filter or empty name will match all Apps.
 	GetApps(ctx context.Context, filter *AppFilter) ([]*App, error)
+
+	// InsertApp inserts an App. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsAlreadyExists if an App by the same name already exists.
 	InsertApp(ctx context.Context, app *App) (*App, error)
+
+	// UpdateApp updates an App's Config. Returns ErrDatastoreEmptyApp when app is nil, and
+	// ErrDatastoreEmptyAppName when app.Name is empty.
+	// Returns ErrAppsNotFound if an App is not found.
+	//TODO are app.Config's fields added to the existing field or do they replace them?
 	UpdateApp(ctx context.Context, app *App) (*App, error)
+
+	// RemoveApp removes the App named appName. Returns ErrDatastoreEmptyAppName if appName is empty.
+	// Returns ErrAppsNotFound if an App is not found.
+	//TODO remove routes automatically? #528
 	RemoveApp(ctx context.Context, appName string) error
 
+	// GetRoute looks up a matching Route for appName and the literal request route routePath.
+	// Returns ErrDatastoreEmptyAppName when appName is empty, and ErrDatastoreEmptyRoutePath when
+	// routePath is empty.
+	// Returns ErrRoutesNotFound when no matching route is found.
+	//TODO should these return ErrAppsNotFound or RoutesNotFound? (or the unused 'ErrRoutesGet')
 	GetRoute(ctx context.Context, appName, routePath string) (*Route, error)
+
+	// GetRoutes gets a slice of Routes, optionally filtered by filter.
 	GetRoutes(ctx context.Context, filter *RouteFilter) (routes []*Route, err error)
-	GetRoutesByApp(ctx context.Context, appName string, filter *RouteFilter) (routes []*Route, err error)
+
+	// InsertRoute inserts a route. Returns ErrDatastoreEmptyRoute when route is nil, and ErrDatastoreEmptyAppName
+	// or ErrDatastoreEmptyRoutePath for empty AppName or Path.
+	// Returns ErrRoutesAlreadyExists if the exact route.Path already exists, or ErrRoutesCreate if a conflicting
+	// route already exists.
 	InsertRoute(ctx context.Context, route *Route) (*Route, error)
+
+	// UpdateRoute updates route's Config and Header fields. Returns ErrDatastoreEmptyRoute when route is nil, and
+	// ErrDatastoreEmptyAppName or ErrDatastoreEmptyRoutePath for empty AppName or Path.
+	//TODO are route's .Config and .Header fields merged with or do they replace existing fields?
 	UpdateRoute(ctx context.Context, route *Route) (*Route, error)
```

Comment:
Similar question to ```UpdateApp```.  Additionally, do empty string route fields indicate 'set field empty', or 'do not update this field`.